cytotable/convert.py

```diff
@@ -245,34 +248,41 @@ def _get_table_chunk_offsets(
 
             return None
 
-    # find chunk offsets from sql statement
     elif sql_stmt is not None:
-        # gather the total rowcount from csv or sqlite data input sources
         with _duckdb_reader() as ddb_reader:
-            rowcount = int(
-                ddb_reader.execute(
-                    # nosec
-                    f"SELECT COUNT(*) FROM ({sql_stmt})"
-                ).fetchone()[0]
-            )
+            sql_query = f"SELECT {page_key} FROM ({sql_stmt}) ORDER BY {page_key}"
+            keys = ddb_reader.execute(sql_query).fetchall()
+            keys = [key[0] for key in keys]
+
+    # Create chunks of keys
+    chunks = []
+    i = 0
+    while i < len(keys):
+        start_key = keys[i]
+        end_index = min(i + chunk_size, len(keys)) - 1
+        end_key = keys[end_index]
+
+        # Ensure non-overlapping by incrementing the start of the next range
+        next_start_index = end_index + 1
+        if next_start_index < len(keys):
+            next_start_key = keys[next_start_index]
+            while next_start_key == end_key and next_start_index + 1 < len(keys):
+                next_start_index += 1
+                next_start_key = keys[next_start_index]
+            chunks.append((start_key, next_start_key - 1))
+        else:
+            chunks.append((start_key, end_key))
 
-    return list(
-        range(
-            0,
-            # gather rowcount from table and use as maximum for range
-            rowcount,
-            # step through using chunk size
-            chunk_size,
-        )
-    )
+        i = next_start_index
+
+    return chunks
 
 
 @python_app
 def _source_chunk_to_parquet(
     source_group_name: str,
     source: Dict[str, Any],
-    chunk_size: int,
-    offset: int,
+    pageset: Tuple[int, int],
```

Comment:
I see that the type here is `Tuple[int, int]`; does that imply that columns identified in `page_keys` are also integers? If so, you might want to add that to the documentation and/or add a check that they are in fact integers.

Reply:
Thanks, good point - as designed `page_keys` need to be numeric (ints or floats). I'll update the type here along with the documentation aspects.